vncviewer/MonitorArrangement.cxx

```diff
@@ -0,0 +1,475 @@
+  x_min = x_max = y_min = y_max = 0;
+
+  for (int i = 0; i < Fl::screen_count(); i++) {
+    Fl::screen_xywh(x, y, w, h, i);
+    
+    top = y;
+    bottom = y + h;
+    left = x;
+    right = x + w;
+    
+    if (top < y_min)
+      y_min = top;
+
+    if (bottom > y_max)
+      y_max = bottom;
+
+    if (left < x_min)
+      x_min = left;
+
+    if (right > x_max)
+      x_max = right;
+  }
+  
+  return std::make_pair(x_max - x_min, y_max - y_min);
+}
+
+std::pair<int, int> MonitorArrangement::offset()
+{
+  double scale = this->scale();
+  std::pair<int, int> size = this->size();
+  std::pair<int, int> origin = this->origin();
+
+  int offset_x = (this->w()/2) - (size.first/2 * scale);
+  int offset_y = (this->h()/2) - (size.second/2 * scale);
+
+  return std::make_pair(offset_x + abs(origin.first)*scale, offset_y + abs(origin.second)*scale);
+}
+
+std::pair<int, int> MonitorArrangement::origin()
+{
+  int x, y, w, h, ox, oy;
+  ox = oy = 0;
+
+  for (int i = 0; i < Fl::screen_count(); i++) {
+    Fl::screen_xywh(x, y, w, h, i);
+
+    if (x < ox)
+      ox = x;
+
+    if (y < oy)
+      oy = y;
+  }
+
+  return std::make_pair(ox, oy);
+}
+
+bool MonitorArrangement::inside(
+  int top_y, int bottom_y, int left_x, int right_x, int x, int y)
+{
+  return (x > left_x) && (x < right_x) && (y > top_y) && (y < bottom_y);
+}
+
+std::string MonitorArrangement::description(int m)
+{
+  assert(m < (int) m_monitors.size());
+  const size_t name_len = 1024;
+  char name[name_len] = {};
+  int bytes_written = get_monitor_name(m, name, name_len);
+
+  int x, y, w, h;
+  Fl::screen_xywh(x, y, w, h, m);
+  std::stringstream ss;
+
+  if (bytes_written > 0)
+    ss << name << " (" << w << "x" << h << ")";
+  else
+    ss << w << "x" << h;
+
+  return ss.str();
+}
+
+int MonitorArrangement::get_monitor_name(int m, char name[], size_t name_len)
+{
+#if defined(WIN32)
+  int x, y, w, h;
+  Fl::screen_xywh(x, y, w, h, m);
+  return win32_get_monitor_name(x, y, w, h, name, name_len);
+
+#elif defined(__APPLE__)
+  CGDisplayCount count;
+  int bytes_written = 0;
+  CGDirectDisplayID displays[16];
+  
+  if (CGGetActiveDisplayList(16, displays, &count) != kCGErrorSuccess)
+    return -1;
+
+  if (count != (unsigned)Fl::screen_count())
+    return -1;
+
+  if (m >= (int)count)
+    return -1;
+
+  // Notice: Here we assume indices to be ordered the same as in FLTK (we rely on that in cocoa.mm as well).
+  CGDirectDisplayID displayID = displays[m];
+
+  CFDictionaryRef info = IODisplayCreateInfoDictionary(
+    /* display = */ CGDisplayIOServicePort(displayID),
+    /* options = */ kIODisplayOnlyPreferredName);
+
+  CFDictionaryRef dict = (CFDictionaryRef) CFDictionaryGetValue(info, CFSTR(kDisplayProductName));
+  CFIndex dict_len = CFDictionaryGetCount(dict);
+
+  if (dict_len > 0) {
+    CFTypeRef * names = new CFTypeRef[dict_len];
+    CFDictionaryGetKeysAndValues(dict, NULL, (const void **) names);
+
+    if ((sizeof(names) / sizeof(*names)) > 0) {
```

Comment:
Won't this always be true? We've already checked `dict_len`.

Reply:
True, my thought behind this was to look whether `CFdictionaryGetKeysAndValues` actually filled the array with something. I realize that this check won't solve that and it has been replaced with another.